### PR TITLE
Add support for CUBE and ROLLUP with composite sets

### LINKED ISSUE #16981: Parser mistakes grouping set specification for ROW in ROLLUP and CUBE
I'm using trino 412 from here: https://hub.docker.com/r/trinodb/trino.

These two queries should be equivalent, given that `ROLLUP` is just syntax sugar for `GROUPING SETS`:

```sql
select a, b, c, count(*)
from (values (1, 1, 1), (1, 2, 2), (1, 2, 2)) t (a, b, c)
group by rollup (a, (b, c))
order by 1, 2, 3;

select a, b, c, count(*)
from (values (1, 1, 1), (1, 2, 2), (1, 2, 2)) t (a, b, c)
group by grouping sets ((a, b, c), a, ())
order by 1, 2, 3;
```

They should both produce:

```
|a  |b  |c  |count|
|---|---|---|-----|
|1  |1  |1  |1    |
|1  |2  |2  |2    |
|1  |   |   |3    |
|   |   |   |3    |
```

But the trino parser seems to think that `(b, c)` is a row value expression, rather than a grouping set specification:

> SQL Error [63]: Query failed (#20230412_114613_00004_whng5): line 3:21: GROUP BY expression must be a column reference: ROW (b, c)

Given that the parser does the right thing for the `GROUPING SETS` syntax (which is the workaround here, of course), I think that's just a parser bug?

`CUBE` is also affected.

### PR DESCRIPTION
**Pull request by martint**
Previously, CUBE and ROLLUP only allowed simple groups composed of single column references. This change makes it possible to specify composite groups such as:

    GROUP BY CUBE ((a,b), c, (d, e, f))

Fixes #16981

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Add support for CUBE and ROLLUP with composite sets ({issue}`16981`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 1]
```diff
@@ -1274,13 +1275,22 @@ private static List<Set<FieldId>> enumerateGroupingSets(GroupingSetAnalysis grou
...
+                    .collect(toImmutableList());
+
+            partialSets.add(sets);
         }
 
-        for (List<FieldId> rollup : groupingSetAnalysis.getRollups()) {
+        for (List<Set<FieldId>> rollup : groupingSetAnalysis.getRollups()) {
             List<Set<FieldId>> sets = IntStream.rangeClosed(0, rollup.size())
-                    .mapToObj(i -> ImmutableSet.copyOf(rollup.subList(0, i)))
+                    .mapToObj(i -> rollup.subList(0, i).stream()
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:1291]: Please rename `i`. Maybe `prefixLength`?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -3975,18 +3973,18 @@ private void checkGroupingSetsCount(GroupBy node)
...
-                        product = element.getExpressions().size() + 1;
-                    }
-                    else if (element instanceof GroupingSets) {
-                        product = ((GroupingSets) element).getSets().size();
+                    else if (element instanceof GroupingSets groupingSets) {
+                        product = switch (groupingSets.getType()) {
+                            case CUBE -> {
+                                int exponent = ((GroupingSets) element).getSets().size();
+                                if (exponent > 30) {
+                                    throw new ArithmeticException();
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3981]: This is preexisting, but wouldn't we rather throw NOT_SUPPORTED?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -4054,38 +4052,20 @@ private GroupingSetAnalysis analyzeGroupBy(QuerySpecification node, Scope scope,
...
-                                            .map(analysis.getColumnReferenceFields()::get)
-                                            .map(ResolvedField::getFieldId)
-                                            .collect(toImmutableSet()))
-                                    .collect(toImmutableList());
-
-                            rollups.add(rollup);
-                        }
-                        else if (groupingElement instanceof GroupingSets) {
-                            List<Set<FieldId>> groupingSets = ((GroupingSets) groupingElement).getSets().stream()
+                        if (groupingElement instanceof GroupingSets element) {
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4055]: `if` is redundant. We're here only if `groupingElement instanceof SimpleGroupBy` is false.

File: core/trino-parser/src/main/java/io/trino/sql/tree/GroupingSets.java
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-parser/src/main/java/io/trino/sql/tree/GroupingSets.java:1]: shallowEquals doesn’t matter for these. It’s only used for expressions.

I’ll fix equals and hash code, though

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 4]
```diff
@@ -1274,13 +1275,22 @@ private static List<Set<FieldId>> enumerateGroupingSets(GroupingSetAnalysis grou
     {
         List<List<Set<FieldId>>> partialSets = new ArrayList<>();
 
-        for (Set<FieldId> cube : groupingSetAnalysis.getCubes()) {
-            partialSets.add(ImmutableList.copyOf(Sets.powerSet(cube)));
+        for (List<Set<FieldId>> cube : groupingSetAnalysis.getCubes()) {
+            Set<Set<Set<FieldId>>> pc = Sets.powerSet(ImmutableSet.copyOf(cube));
```
[the remark this answers is not in the document]
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:1279]: Oops, leftover from an experiment

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3981]: Yes. I'll take care of it later in a separate PR.

**Pull request merged**